monix-catnap/shared/src/main/scala/monix/catnap/MVar.scala

```diff
@@ -200,175 +199,113 @@ object MVar {
+    def empty[A](ps: PaddingStrategy = NoPadding): F[MVar[F, A]] =
+      MVar.empty(ps)(F)
   }
 
-  private trait Impl[F[_], A] { self: GenericVar[A, F[Unit]] =>
-    implicit def F: Async[F]
-    implicit def cs: ContextShift[F]
+  private final class AsyncImpl[F[_], A](initial: Option[A], ps: PaddingStrategy)(
+    implicit val F: Async[F])
+    extends GenericVar[A, F[Unit]](initial, ps) {
 
-    protected def create[T](k: (Either[Throwable, T] => Unit) => F[Unit]): F[T]
+    override protected def makeCancelable(f: Id => Unit, id: Id): F[Unit] =
+      F.delay(f(id))
+    override protected def emptyCancelable: F[Unit] =
+      F.unit
 
     def swap(newValue: A): F[A] =
-      F.flatMap(take) { oldValue =>
-        F.map(put(newValue))(_ => oldValue)
+      F.uncancelable { poll =>
+        F.flatMap(poll(take))(F.as(put(newValue), _))
       }
 
+
+    def modify_(f: A => F[A]): F[Unit] =
+      modify(a => F.map(f(a))((_, ())))
+
     def use[B](f: A => F[B]): F[B] =
       modify(a => F.map(f(a))((a, _)))
 
     def modify[B](f: A => F[(A, B)]): F[B] =
-      F.flatMap(take) { a =>
-        F.flatMap(F.onError(f(a)) { case _ => put(a) }) {
-          case (newA, b) =>
-            F.as(put(newA), b)
+      F.uncancelable { poll =>
+        F.flatMap(poll(take)) { oldA =>
+          F.flatMap(poll(F.onCancel(f(oldA), put(oldA)))) { case (a, b) =>
+            F.as(put(a), b)
+          }
         }
       }
 
-    def modify_(f: A => F[A]): F[Unit] =
-      modify(a => F.map(f(a))((_, ())))
-
-    final def isEmpty: F[Boolean] =
+    def isEmpty: F[Boolean] =
       F.delay(unsafeIsEmpty())
-    final def tryPut(a: A): F[Boolean] =
+    def tryPut(a: A): F[Boolean] =
       F.delay(unsafeTryPut(a))
-    final def tryTake: F[Option[A]] =
+    def tryTake: F[Option[A]] =
       F.delay(unsafeTryTake())
-    final def tryRead: F[Option[A]] =
+    def tryRead: F[Option[A]] =
       F.delay(unsafeTryRead())
 
-    final def put(a: A): F[Unit] =
-      F.suspend {
+    def put(a: A): F[Unit] =
+      F.defer {
         if (unsafeTryPut(a))
           F.unit
         else
           F.flatMap(AsyncUtils.cancelable[F, Unit](unsafePut(a, _)))(bindFork)
       }
 
-    final def take: F[A] =
-      F.suspend[A] {
+    def take: F[A] =
+      F.defer[A] {
         unsafeTryTake() match {
           case Some(a) => F.pure(a)
           case None =>
             F.flatMap(AsyncUtils.cancelable[F, A](unsafeTake))(bindForkA.asInstanceOf[A => F[A]])
         }
       }
 
-    final def read: F[A] =
+    def read: F[A] =
       AsyncUtils.cancelable { cb =>
         unsafeRead(cb)
       }
 
+    private[this] val shift = F.cede
+
     private[this] val bindFork: (Unit => F[Unit]) = {
-      val shift = cs.shift
       _ => shift
     }
 
     private[this] val bindForkA: (Any => F[Any]) = {
-      val shift = cs.shift
       x => F.map(shift)(_ => x)
     }
   }
-
-  private final class AsyncImpl[F[_], A](initial: Option[A], ps: PaddingStrategy)(
```

Comment:
Since CE3 is dropping whole "uncancelable Async" thing, and our Task is already always old Concurrent, new Async, I think we should follow suit. The original rationale was to allow users to not require `Concurrent` if they just need a tiny bit of semantic blocking, but since Concurrent was a control flow thing, it was required quite pervasively anyway.